File: java/shared/java/org/signal/libsignal/protocol/incrementalmac/ChunkSizeChoice.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.protocol.incrementalmac;

import org.signal.libsignal.internal.Native;

public abstract class ChunkSizeChoice {

    public abstract int getSizeInBytes();

    public static ChunkSizeChoice everyNthByte(int n) {
        return new EveryN(n);
    }

    public static ChunkSizeChoice inferChunkSize(int dataSize) {
        return new ChunksOf(dataSize);
    }

    private static final class EveryN extends ChunkSizeChoice {
        private int n;

        private EveryN(int n) {
            this.n = n;
        }

        public int getSizeInBytes() {
            return this.n;
        }
    }

    private static final class ChunksOf extends ChunkSizeChoice {
        private int dataSize;

        private ChunksOf(int dataSize) {
            this.dataSize = dataSize;
        }

        public int getSizeInBytes() {
            return Native.IncrementalMac_CalculateChunkSize(this.dataSize);
        }
    }
}

File: java/shared/java/org/signal/libsignal/protocol/incrementalmac/ChunkSizeChoice.kt
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
package org.signal.libsignal.protocol.incrementalmac

import org.signal.libsignal.internal.Native

public abstract class ChunkSizeChoice {
  public abstract val sizeInBytes: Int

  public companion object {
    @JvmStatic
    public fun everyNthByte(n: Int): ChunkSizeChoice = EveryN(n)

    @JvmStatic
    public fun inferChunkSize(dataSize: Int): ChunkSizeChoice = ChunksOf(dataSize)
  }
}

internal final data class EveryN(
  val n: Int,
) : ChunkSizeChoice() {
  override val sizeInBytes: Int = n
}

internal final data class ChunksOf(
  val dataSize: Int,
) : ChunkSizeChoice() {
  override val sizeInBytes: Int = Native.IncrementalMac_CalculateChunkSize(this.dataSize)
}
